app/models/manageiq/providers/inventory/persister/builder/shared.rb

```diff
@@ -72,9 +72,31 @@ def hardwares
     end
 
     def operating_systems
+      custom_save_block = lambda do |_ems, inventory_collection|
+        vms_and_templates_ids = inventory_collection.data.map { |os| os.vm_or_template&.id }.compact
+
+        vms_and_templates_index = VmOrTemplate.where(:id => vms_and_templates_ids).index_by(&:id)
+        operating_systems_index = OperatingSystem.where(:vm_or_template_id => vms_and_templates_ids).index_by(&:vm_or_template_id)
```

Comment:
This feels like it should be a joined thing...

```ruby
VmOrTemplate.includes(:operating_system).where(:id => vms_and_templates_ids).index_by(&:id)
```

Reply:
Done